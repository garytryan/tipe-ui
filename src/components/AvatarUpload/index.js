export { default } from './AvatarUpload.vue'
