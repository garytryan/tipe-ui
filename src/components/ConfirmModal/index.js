export { default } from './ConfirmModal.vue'
