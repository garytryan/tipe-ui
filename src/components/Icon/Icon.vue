<template>
  <svg
    :data-tipe-ui="$options.name"
    :style="styleObject"
    :viewBox="src.viewBox"
    :color="color"
  >
    <use :xlink:href="src.id" />
  </svg>
</template>

<script>
import vueTypes from 'vue-types'
import iconLoader from '@/libs/iconLoader'

export default {
  name: 'TipeIcon',
  props: {
    icon: vueTypes.string.def('plus'),
    height: vueTypes.string.def('1rem'),
    width: vueTypes.string.def('1rem'),
    color: vueTypes.string.def('#000'),
    __iconLoader__: vueTypes.func.def(iconLoader)
  },
  computed: {
    src() {
      const { viewBox, id } = this.__iconLoader__(this.icon)
      return {
        viewBox,
        id: `#${id}`
      }
    },
    styleObject() {
      return {
        height: this.height,
        width: this.width
      }
    }
  }
}
</script>
