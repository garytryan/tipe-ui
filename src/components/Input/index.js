export SimpleInput from './SimpleInput.vue'
