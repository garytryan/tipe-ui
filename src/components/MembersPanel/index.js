export TableRow from './TableRow'
export TipeTable from './Table'
export AddMember from './AddMember'
export { default } from './MembersPanel'
