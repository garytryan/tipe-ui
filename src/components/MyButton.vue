<template>
  <button
    @click="onClick"
  >
    <slot />
  </button>
</template>

<script>
export default {
  name: 'MyButton',

  methods: {
    onClick() {
      this.$emit('click')
    }
  }
}
</script>

<style scoped>
button {
  border: '1px solid #eee';
  border-radius: 3;
  background-color: '#FFFFFF';
  cursor: 'pointer';
  font-size: 15;
  padding: '3px 10px';
  margin: 10;
}
</style>
