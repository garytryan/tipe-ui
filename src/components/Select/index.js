export { default } from './Select.vue'
export TipeSelectButton from './SelectButton.vue'
export TipeSelectUnderline from './Select.vue'
