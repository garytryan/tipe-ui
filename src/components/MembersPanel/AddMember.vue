<template>
  <div class="grid">
    <div class="close"><icon icon="close"/></div>
    <div class="content">
      <p class="title">Add Member</p>
      <avatar-upload />
      <form @submit="onSubmit">
        <tipe-input 
          text-label="Name" 
          input-type="text" 
          class="name"
          :value="nameValue"
          :validity="nameValidity"
          text-placeholder="Type your name here"
          @change="onChangeName"/>
        <tipe-input 
          text-label="Email" 
          class="email"
          :value="emailValue"
          input-type="email" 
          :validity="emailValidity"
          text-placeholder="Type your email here"
          @change="onChangeEmail"/>
        <tipe-select 
          :options="options" 
          :value="roleValue" 
          class="role"
          text-label="Role"
          @change="onChangeRole" />
        <div class="seperator"><seperator /></div>

        <div class="add-button">
          <tipe-button color="purple" size="full">ADD MEMBER</tipe-button>
        </div>
      </form>
    </div>
  </div>
</template>

<script>
import { createManyMocks, user } from '../../mocks'
import Icon from '../Icon'
import TipeButton from '../Button'
import Seperator from '../Seperator'
import AvatarUpload from '../AvatarUpload'
import TipeInput from '../Input'
import TipeSelect from '../Select'

export default {
  name: 'AddMember',
  components: {
    Icon,
    TipeButton,
    Seperator,
    AvatarUpload,
    TipeInput,
    TipeSelect
  },
  data() {
    return {
      members: createManyMocks(user, 12),
      options: [
        { label: 'Owner', value: 'owner' },
        { label: 'Member', value: 'member' },
        { label: 'Manager', value: 'manager' }
      ],
      emailValue: '',
      roleValue: 'member',
      nameValue: ''
    }
  },
  methods: {
    onChangeEmail(val) {
      this.emailValue = val
    },
    onChangeRole(val) {
      this.roleValue = val
    },
    onChangeName(val) {
      this.nameValue = val
    },
    onSubmit() {}
  },
  computed: {
    emailValidity() {
      var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
      let valid = re.test(this.emailValue)
      let status = valid === true ? 'success' : 'error'
      let message =
        status === 'error' ? 'Please enter a valid email address' : ''
      return {
        valid,
        status,
        message
      }
    },
    nameValidity() {
      let valid = this.nameValue.trim() !== ''
      let status = valid === true ? 'success' : 'error'
      let message = status === 'error' ? 'Please enter a valid name' : ''
      return {
        valid: this.nameValue !== '',
        status,
        message
      }
    },
    canSubmit() {
      return this.emailValidity.valid && this.nameValidity.valid
    }
  }
}
</script>


<style lang='postcss' scoped>
.grid {
  display: grid;
  background-color: #fff;
  border-radius: 1rem;
  width: 550px;
  max-height: 100%;
  grid-template-columns: 3rem auto 3rem;
  grid-template-rows: 3rem auto;
  grid-template-areas:
    '. . close'
    'content content content';
}

.close {
  background-color: var(--gray-light);
  grid-area: close;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 0 1rem 0 0;
}

.add-button {
  display: flex;
  width: 100%;
  align-items: center;
  justify-content: center;
  grid-area: button;
}

.seperator {
  grid-area: seperator;
}

.name {
  grid-area: nameInput;
}

.email {
  grid-area: emailInput;
}

.role {
  grid-area: roleInput;
}

.content {
  grid-area: content;
  display: grid;
  width: 100%;
  align-content: start;
  grid-row-gap: 2rem;

  & .title {
    color: var(--darkPurple);
    font-size: 1.5rem;
    font-weight: bold;
    margin: 0;
    justify-self: center;
  }

  & form {
    display: grid;
    grid-row-gap: 1.25rem;
    grid-template-rows: 1fr 1fr 1fr 0 6rem;
    grid-template-columns: 3rem auto 3rem;
    grid-template-areas:
      '. nameInput .'
      '. emailInput .'
      ' . roleInput .'
      'seperator seperator seperator'
      '. button .';
  }
}
</style>
