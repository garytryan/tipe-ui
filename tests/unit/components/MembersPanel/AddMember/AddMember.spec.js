import { shallowMount } from '@vue/test-utils'
import { AddMember } from '@/components/MembersPanel'
import { createRenderer } from 'vue-server-renderer'

import Icon from '@/components/Icon'
import TipeButton from '@/components/Button'
import Seperator from '@/components/Seperator'
import AvatarUpload from '@/components/AvatarUpload'
import TipeInput from '@/components/Input'
import TipeSelect from '@/components/Select'

describe('AddMember.vue', () => {
  it('renders', () => {
    const renderer = createRenderer()
    const wrapper = shallowMount(AddMember)
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })
  it('should render the Icon component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(Icon)).toBe(true)
  })
  it('should render the TipeButton component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(TipeButton)).toBe(true)
  })
  it('should render the Seperator component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(Seperator)).toBe(true)
  })
  it('should render the AvatarUpload component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(AvatarUpload)).toBe(true)
  })
  it('should render the TipeInput component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(TipeInput)).toBe(true)
  })
  it('should render the TipeSelect component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(TipeSelect)).toBe(true)
  })
  it('should have the grid class', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.classes()).toContain('grid')
  })
  it('should invalidate the email input', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({ emailValue: { value: '454', changed: true } })
    expect(wrapper.vm.emailValidity).toEqual({
      valid: false,
      status: 'error',
      message: 'Please enter a valid email address'
    })
  })
  it('should validate the email input', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({ emailValue: { value: 'olivia@tipe.io', changed: true } })
    expect(wrapper.vm.emailValidity).toEqual({
      valid: true,
      status: 'success',
      message: ''
    })
  })
  it('should invalidate the name input', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({ nameValue: { value: '', changed: true } })
    expect(wrapper.vm.nameValidity).toEqual({
      valid: false,
      status: 'error',
      message: 'Please enter a valid name'
    })
  })
  it('should validate the name input', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({ nameValue: { value: 'Olivia Oddo', changed: true } })
    expect(wrapper.vm.nameValidity).toEqual({
      valid: true,
      status: 'success',
      message: ''
    })
  })
  it('should not be able to submit the form', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({
      emailValue: {
        value: '',
        changed: true
      },
      roleValue: 'member',
      nameValue: { value: '', changed: true }
    })
    expect(wrapper.vm.canSubmit).toBe(false)
  })
  it('should not be able to submit the form with invalid email', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({
      emailValue: {
        value: 'dsfdsfs',
        changed: true
      },
      roleValue: 'member',
      nameValue: { value: 'Olivia', changed: true }
    })
    expect(wrapper.vm.canSubmit).toBe(false)
  })
  it('should not be able to submit the form with invalid name', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({
      emailValue: {
        value: 'olivia.oddo@gmail.com',
        changed: true
      },
      roleValue: 'member',
      nameValue: { value: '', changed: true }
    })
    expect(wrapper.vm.canSubmit).toBe(false)
  })
  it('should be able to submit the form', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({
      emailValue: {
        value: 'olivia.oddo@gmail.com',
        changed: true
      },
      roleValue: 'member',
      nameValue: { value: 'Olivia', changed: true }
    })
    expect(wrapper.vm.canSubmit).toBe(true)
  })
  it('should reset the form', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({
      emailValue: {
        value: 'olivia.oddo@gmail.com',
        changed: true
      },
      roleValue: 'member',
      nameValue: { value: 'Olivia', changed: true }
    })
    wrapper.vm.resetForm()
    expect(wrapper.vm.roleValue).toBe('member')
    expect(wrapper.vm.nameValue).toEqual({ value: '', changed: false })
    expect(wrapper.vm.emailValue).toEqual({
      value: '',
      changed: false
    })
  })
  it('should update the name value', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.vm.onChangeName('Olivia')
    expect(wrapper.vm.nameValue).toEqual({ value: 'Olivia', changed: true })
  })
  it('should update the email value', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.vm.onChangeEmail('olivia.oddo@gmail.com')
    expect(wrapper.vm.emailValue).toEqual({
      value: 'olivia.oddo@gmail.com',
      changed: true
    })
  })
  it('should update the role value', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.vm.onChangeRole('owner')
    expect(wrapper.vm.roleValue).toEqual('owner')
  })
})
