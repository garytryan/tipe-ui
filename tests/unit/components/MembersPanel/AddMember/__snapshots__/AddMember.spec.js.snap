// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMember.vue renders 1`] = `
<div data-tipe-ui="AddMember" class="grid">
  <div class="close">
    <!---->
  </div>
  <div class="content">
    <p class="title">Add Member</p>
    <!---->
    <form>
      <!---->
      <!---->
      <!---->
      <div class="seperator">
        <!---->
      </div>
      <div class="add-button">
        <!---->
      </div>
    </form>
  </div>
</div>
`;
